python/pyarrow/dataset.py

```diff
@@ -232,104 +228,261 @@ def _ensure_format(obj):
+    elif not isinstance(filesystem, FileSystem):
         raise TypeError(
-            "Expected a path-like or DatasetFactory, got {}".format(type(src))
+            '`filesystem` argument must be a FileSystem instance or a valid '
+            'file system URI'
         )
 
+    # retrieve the file descriptor if it is available already
+    if file_info is None:
+        file_info = filesystem.get_file_info([path])[0]
+
+    # depending on the path type either return with a recursive
+    # directory selector or as a list containing a single file
+    if file_info.type == FileType.Directory:
+        paths_or_selector = FileSelector(path, recursive=True)
+    elif file_info.type == FileType.File:
+        paths_or_selector = [path]
+    else:
+        raise FileNotFoundError(path)
+
+    return filesystem, paths_or_selector
 
-def dataset(paths_or_factories, filesystem=None, partitioning=None,
-            format=None, schema=None):
+
+def _filesystem_dataset(source, schema=None, filesystem=None,
+                        partitioning=None, format=None,
+                        partition_base_dir=None, exclude_invalid_files=None,
+                        ignore_prefixes=None):
+    """
+    Create a FileSystemDataset which can be used to build a Dataset.
+
+    Parameters are documented in the dataset function.
+
+    Returns
+    -------
+    FileSystemDataset
+    """
+    format = _ensure_format(format or 'parquet')
+    partitioning = _ensure_partitioning(partitioning)
+
+    if isinstance(source, (list, tuple)):
+        fs, paths_or_selector = _ensure_multiple_sources(source, filesystem)
+    else:
+        fs, paths_or_selector = _ensure_single_source(source, filesystem)
+
+    options = FileSystemFactoryOptions(
+        partitioning=partitioning,
+        partition_base_dir=partition_base_dir,
+        exclude_invalid_files=exclude_invalid_files,
+        ignore_prefixes=ignore_prefixes
+    )
+    factory = FileSystemDatasetFactory(fs, paths_or_selector, format, options)
+
+    return factory.finish(schema)
+
+
+def _union_dataset(source, schema=None, **kwargs):
+    if any(v is not None for v in kwargs.values()):
+        raise ValueError(
+            "When passing a list of Datasets, you cannot pass any additional "
+            "arguments"
+        )
+
+    if schema is None:
+        # unify the children datasets' schemas
+        schema = pa.unify_schemas([ds.schema for ds in source])
+
+    # create datasets with the requested schema
+    children = [ds.replace_schema(schema) for ds in source]
+
+    return UnionDataset(schema, children)
+
+
+def dataset(source, schema=None, format=None, filesystem=None,
+            partitioning=None, partition_base_dir=None,
+            exclude_invalid_files=None, ignore_prefixes=None):
     """
     Open a dataset.
 
     Parameters
     ----------
-    paths_or_factories : path or list of paths or factory or list of factories
+    source : path, list of paths, dataset or list of datasets
```

Comment:
```suggestion
    source : path, list of paths, dataset, list of datasets, factory, or list of factories
```

Not sure what's legal in docstrings like this, but:
```suggestion
    source : SingleOrList[path], SingleOrList[dataset], or SingleOrList[factory]
```

Reply:
I'm not sure that it currently works with factories. I'm checking it before changing the docstring.

Reply:
I don't think it needs to work with a factory? (If you are using a factory, you can just call the finish method yourself to get a Dataset)

> SingleOrList[path], SingleOrList[dataset], or SingleOrList[factory]

I think the "plain english" is easier to understand

Reply:
After taking a try, I think the high level dataset function should not support factory objects. If someone creates factory objects it's unlikely that she/he would pass it to dataset rather than manually calling finish.